lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs

```diff
@@ -485,3 +499,75 @@ impl Error {
         }
     }
 }
+
+
+
+// ==================
+// === Controller ===
+// ==================
+
+/// Controls the shader compiler. While the shader compiler will be thrown away if context is lost,
+/// the controller's state is persistent.
+#[derive(Debug, Clone, CloneRef, Default)]
+pub struct Controller {
+    rc: Rc<RefCell<ControllerData>>,
+}
+
+#[derive(Default, Derivative)]
```

Comment:
1. The docs doesnt explain what is the purpose this struct exists. I suspect its needed for performance tracking, but I know it only because it is part of this PR, I would never know it when reading the code - this should be explained in the docs.
2. Could you use our callback manager instead here? There are a lot of different callback registries defined, like `callback::registry::CopyMut1` - I believe you'd find something with exactly this functionality there. However, if you came to a conclusion that no implementation there has matching performance to this one, I'd love to know why, because these callback registries were meant to be lightweight and used widely.